champions/src/main/java/me/mykindos/betterpvp/champions/champions/skills/skills/brute/passives/Juggle.java

```diff
@@ -0,0 +1,272 @@
+import me.mykindos.betterpvp.core.listener.BPvPListener;
+import me.mykindos.betterpvp.core.scheduler.BPVPTask;
+import me.mykindos.betterpvp.core.scheduler.TaskScheduler;
+import me.mykindos.betterpvp.core.utilities.UtilBlock;
+import me.mykindos.betterpvp.core.utilities.UtilEntity;
+import me.mykindos.betterpvp.core.utilities.UtilMessage;
+import me.mykindos.betterpvp.core.utilities.UtilVelocity;
+import me.mykindos.betterpvp.core.utilities.math.VelocityData;
+import org.bukkit.*;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.entity.Player;
+import org.bukkit.event.EventHandler;
+import org.bukkit.event.EventPriority;
+import org.bukkit.event.Listener;
+import org.bukkit.event.entity.EntityDamageEvent;
+import org.bukkit.util.Vector;
+
+import java.util.*;
+
+@Singleton
+@BPvPListener
+public class Juggle extends Skill implements PassiveSkill, OffensiveSkill, TeamSkill, CrowdControlSkill, Listener {
+
+    private double timeBetweenCharges;
+    private double timeBetweenChargesDecreasePerLevel;
+    private int baseCharges;
+    private int chargesIncreasePerLevel;
+
+    private double enemyVelocity;
+    private double enemyYMax;
+    private double enemyYAdd;
+    private double enemyYSet;
+
+    private double allyVelocity;
+    private double allyYMax;
+    private double allyYAdd;
+    private double allyYSet;
+
+    private double fallDamageLimit;
+
+    private final WeakHashMap<UUID, JuggleData> data = new WeakHashMap<>();
+
+    private final TaskScheduler taskScheduler;
+
+    @Inject
+    public Juggle(Champions champions, ChampionsManager championsManager, TaskScheduler taskScheduler) {
+        super(champions, championsManager);
+        this.taskScheduler = taskScheduler;
+    }
+
+    @Override
+    public String getName() {
+        return "Juggle";
+    }
+
+    @Override
+    public String[] getDescription(int level) {
+        return new String[]{
+                "While you're airborne, strike an ally",
+                "or enemy to briefly send them upward.",
+                "",
+                "Store up to " + getValueString(this::getMaxCharges, level) + " charges",
+                "",
+                "Gain a charge every: " + getValueString(this::getTimeBetweenCharges, level) + " seconds"
+        };
+    }
+
+    @Override
+    public Role getClassType() {
+        return Role.BRUTE;
+    }
+
+    @Override
+    public SkillType getType() {
+        return SkillType.PASSIVE_B;
+    }
+
+    public int getMaxCharges(int level) {
+        return baseCharges + ((level - 1) * chargesIncreasePerLevel);
+    }
+
+    public double getTimeBetweenCharges(int level) {
+        return timeBetweenCharges - ((level - 1) * timeBetweenChargesDecreasePerLevel);
+    }
+
+    @EventHandler(priority = EventPriority.LOWEST)
+    public void onHitAlly(PreCustomDamageEvent event) {
+        CustomDamageEvent cde = event.getCustomDamageEvent();
+
+        if (!(cde.getDamager() instanceof Player player)) return;
+
```

Comment:
You should check the damage cause here, no need to execute this for anything other than EntityAttack I assume?